File: 백준/Gold/11049. 행렬 곱셈 순서/행렬 곱셈 순서.py
import sys
input = sys.stdin.readline

case_num = int(input())
matrix = []
for _ in range(case_num):
    tmp = list(map(int, input().split()))
    matrix.append(tmp)
    
#print(matrix)
sol = [[0] * case_num for _ in range(case_num)]
for len in range(1, case_num): # 구간의 범위 
    for i in range(case_num-len): # 시작 지점
        j = i + len
        sol[i][j] = 2**31
        for k in range(i, j): # k를 기준으로 최솟값 갱신
            sol[i][j] = min(sol[i][j], sol[i][k] + sol[k+1][j] + matrix[i][0] * matrix[k][1] * matrix[j][1])
            
print(sol[0][-1])

File: 백준/Gold/11049. 행렬 곱셈 순서/행렬 곱셈 순서.cc
#include <iostream>
#include <algorithm>
using namespace std;
#define MAX 1000000000

int main() {
    ios::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);
    
    int case_num;
    cin >> case_num;
    int matrix[502][2];
    int sol[502][502];

    for(int i = 1; i <= case_num; i++)
        cin >> matrix[i][0] >> matrix[i][1];

    for(int len = 1; len < case_num; len++) { // 구간의 범위 -> 곱할 행렬의 개수
        for(int j = 1; len+j <= case_num; j++) { // 연산 시작 지점
            sol[j][len+j] = MAX;
            for(int k = j; k <= len+j; k++) { // 기준점
                sol[j][len+j] = min(sol[j][len+j], sol[j][k] + sol[k+1][len+j] + matrix[j][0] * matrix[k][1] * matrix[len+j][1]);
            }
        }
    }
    
    cout << sol[1][case_num] << endl;

    return 0;
}